### PR TITLE
xcode support

### PR DESCRIPTION
**Pull request by mkareta**
Added minimal Xcode support:

- Use multi configuration compile options to support IDEs 

- Xcode defines `SDK_VERSION` as versions of macOS SDK, which conflicts with the osquery SDK. 
Moreover it doesn't make sense to define "release" style `OSQUERY_BUILD_SDK_VERSION` when building through an IDE.

Tested on High Sierra & Mojve using Xcode 9.4

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -486,8 +486,10 @@ if(OSQUERY_BUILD_VERSION STREQUAL "")
   WARNING_LOG("Setting the version to: 1.0.0-unknown (beware!)")
   set(OSQUERY_BUILD_VERSION "1.0.0-unknown")
 endif()
-if(DEFINED ENV{SDK_VERSION})
-  set(OSQUERY_BUILD_SDK_VERSION "${ENV{SDK_VERSION}}")
+
+# SDK_VERSION is defined by XCode as current osx sdk version
```
Comment (obelisk) [inline CMakeLists.txt]: macOS, not OS X

**Comment by mkareta** [comment]
ok to test

File: CMakeLists.txt [hunk 2]
```diff
@@ -288,34 +272,71 @@ elseif(DEFINED ENV{SANITIZE})
...
+# XCODE_ATTRIBUTE_* for specific target 
+# all global definitions/compile options will be inherited 
+# using xcode $(inherited) variable
+if(CMAKE_GENERATOR STREQUAL Xcode)
+  set(CMAKE_XCODE_ATTRIBUTE_GCC_PREPROCESSOR_DEFINITIONS[variant=Debug] "DEBUG=1")
+  set(CMAKE_XCODE_ATTRIBUTE_GCC_PREPROCESSOR_DEFINITIONS[variant=Release] "NDEBUG=1")
+  set_property(DIRECTORY APPEND PROPERTY COMPILE_DEFINITIONS
+    "$(inherited)"
+  )
+  # set_property(DIRECTORY APPEND PROPERTY COMPILE_OPTIONS
```
Comment (obelisk) [inline CMakeLists.txt]: Nix

File: CMakeLists.txt [hunk 3]
```diff
@@ -288,34 +272,71 @@ elseif(DEFINED ENV{SANITIZE})
...
+  #   "$(inherited)"
+  # )
+else()
+  if(CMAKE_BUILD_TYPE STREQUAL "DEBUG")
+    add_definitions(-DDEBUG)
   else()
-    add_compile_options(-Os)
```
Comment (obelisk) [inline CMakeLists.txt]: This is readded below on L311 correct?

File: CMakeLists.txt [hunk 4]
```diff
@@ -288,34 +272,71 @@ elseif(DEFINED ENV{SANITIZE})
...
+# Do not enable fortify with clang: http://llvm.org/bugs/show_bug.cgi?id=16821
+# Visual Studio does not support generation of PIC code
+if (POSIX)
+  # TSAN and MSAN do not like PIE/PIC on Linux 4.1.
+  # To be safe, only include them when building 'release' outputs.
+  add_compile_options(
+    "$<$<CONFIG:RELEASE>:-g>"
+    "$<$<CONFIG:RELEASE>:-fno-limit-debug-info>"
+    "$<$<CONFIG:RELEASE>:-fPIE>"
+    "$<$<CONFIG:RELEASE>:-fpie>"
```
Comment (obelisk) [inline CMakeLists.txt]: Are these two different?

Comment (mkareta) [reply CMakeLists.txt]: who knows :)

**Comment by obelisk** [comment]
When I run this I get a lot of race conditions. Specifically for the osquery_amalgamation files. I get build failures because they don't exist and it seems to happen because it's trying to build something that uses it before generation.

File: CMakeLists.txt [hunk 5]
```diff
@@ -286,9 +286,9 @@ if(CMAKE_GENERATOR STREQUAL Xcode)
   set_property(DIRECTORY APPEND PROPERTY COMPILE_DEFINITIONS
     "$(inherited)"
   )
-  # set_property(DIRECTORY APPEND PROPERTY COMPILE_OPTIONS
-  #   "$(inherited)"
-  # )
+  set_property(DIRECTORY APPEND PROPERTY COMPILE_OPTIONS
```
Comment (obelisk) [inline CMakeLists.txt]: Can this be merged with the above block?
```
set_property(DIRECTORY APPEND PROPERTY COMPILE_DEFINITIONS COMPILE_OPTIONS
     "$(inherited)"
)
```
I.e Does that work?

Comment (mkareta) [reply CMakeLists.txt]: Unfortunately you can specify only one property
https://cmake.org/cmake/help/v3.10/command/set_property.html

**Pull request merged**